# Machine Learning para prever Resistência a Antibióticos


## Introdução 
A resistência aos antibióticos tornou-se uma preocupação global de saúde pública. As bactérias estão desenvolvendo resistência ao arsenal de antibióticos prescritos, resultando em variações que estão desenvolvendo resistência a múltiplos medicamentos. Atualmente, os médicos e as clínicas realizam ensaios tradicionais baseados em cultura (ou seja, testando se um medicamento mataria as bactérias) para determinar a resistência aos antibióticos em estirpes bacterianas.

Alternativamente, as clínicas também podem optar por sequenciar as cepas. Essas sequências podem então ser analisadas para prever se apresentará resistência a um determinado medicamento. Existem diferentes maneiras de realizar a análise e o aprendizado de máquina é uma delas. 

Neste projeto processaremos sequências completas do genoma de cepas de E. coli disponíveis publicamente para construir um modelo de Machine Learning que será usado para prever a Resistência (R) e a Suscetibilidade (S) para cada cepa. As cepas já foram testadas em laboratório, então poderemos mais tarde comparar as previsões feitas por nosso modelo de Machine Learning com os resultados dos ensaios tradicionais baseados em cultura, a fim de determinar o desempenho do modelo que criarmos.

### O que é resistência a antibióticos?

A resistência aos antibióticos é um fenômeno pelo qual bactérias e outros microrganismos desenvolvem a capacidade de sobreviver e se proliferar na presença de antibióticos que, anteriormente, seriam capazes de matá-los ou inibir seu crescimento.

Esse processo de resistência ocorre através de mutações genéticas e/ou pela aquisição de genes de resistência de outras bactérias, o que torna os tratamentos com antibióticos menos eficazes e pode levar a infecções difíceis de tratar.


#### Como a Resistência aos Antibióticos se Desenvolve


**Mutação Genética:** Bactérias podem sofrer mutações espontâneas em seu DNA que lhes conferem resistência aos antibióticos. Essas mutações podem ocorrer de forma natural e aleatória.

**Transferência Horizontal de Genes:** Bactérias podem adquirir genes de resistência de outras bactérias através de mecanismos como:


- Conjugação: Transferência de material genético diretamente de uma bactéria para outra através de contato físico.
- Transformação: Captação de DNA livre do ambiente e incorporação no genoma bacteriano.
- Transdução: Transferência de genes de resistência mediada por vírus bacteriófagos.
- Seleção Natural: Quando uma população bacteriana é exposta a um antibiótico, as bactérias sensíveis ao medicamento são eliminadas, enquanto as resistentes sobrevivem e se multiplicam. Esse processo resulta em uma população predominantemente resistente.

#### Consequências da Resistência aos Antibióticos


- Dificuldade no Tratamento de Infecções: Infecções que antes eram facilmente tratáveis com antibióticos comuns podem se tornar difíceis ou impossíveis de tratar.
- Maior Tempo de Doença e Mortalidade: Infecções resistentes podem levar a doenças prolongadas, complicações mais graves e um aumento na taxa de mortalidade.
- Custos de Saúde Elevados: Tratamentos de infecções resistentes são geralmente mais caros, exigindo medicamentos de segunda ou terceira linha, internações prolongadas e cuidados adicionais.
- Risco de Epidemias: A propagação de bactérias resistentes pode levar a surtos e epidemias, dificultando o controle de doenças infecciosas.


Exemplos de Bactérias Resistentes

- MRSA (Staphylococcus aureus resistente à meticilina): Uma cepa de Staphylococcus aureus que é resistente à meticilina e outros antibióticos beta-lactâmicos.
- VRE (Enterococos resistentes à vancomicina): Enterococos que adquiriram resistência à vancomicina, um antibiótico frequentemente usado como último recurso.
- ESBL (Enterobactérias produtoras de beta-lactamase de espectro estendido): Bactérias que produzem enzimas capazes de destruir uma ampla gama de antibióticos beta-lactâmicos.


Prevenção e Controle

- Uso Racional de Antibióticos: Evitar o uso excessivo e desnecessário de antibióticos tanto na medicina quanto na agricultura.
- Higiene e Saneamento: Melhorar as práticas de higiene e saneamento para prevenir a propagação de infecções.
- Vacinação: Usar vacinas para prevenir infecções bacterianas e reduzir a necessidade de antibióticos.
- Desenvolvimento de Novos Antibióticos: Investir em pesquisa e desenvolvimento de novos antibióticos e tratamentos alternativos.
- Monitoramento e Vigilância: Implementar sistemas eficazes de monitoramento e vigilância para detectar e controlar surtos de resistência.
- A resistência aos antibióticos é um dos maiores desafios de saúde pública do século XXI, exigindo esforços coordenados a nível global para sua contenção e manejo.



## Desenvolvimento

### Pacotes



In [36]:
# Imports
import pandas as pd
import numpy as np
from functools import reduce

### Dados

#### Dataset 1

Os dados usados neste projeto foram gerados a partir do estudo chamado: Prediction of antibiotic resistance in Escherichia coli from large-scale pan-genome data.

Publicação:  Moradigaravand  D,  Palm  M,  Farewell  A,  Mustonen  V,  Warringer  J,  PartsL (2018). PLoS Comput Biol 14(12): e1006258. https://doi.org/10.1371/journal.pcbi.1006258

Página do Github: https://github.com/DaneshMoradigaravand/PanPred

Neste projeto vamos trabalhar com 3 datasets:

a) **Dataset 1:** Metadados com ano de isolamento e resultados dos Testes de Suscetibilidade Antimicrobiana.

Descrição: Este dataset contém informações sobre isolados bacterianos, incluindo o ano em que foram isolados e os resultados dos Testes de Suscetibilidade Antimicrobiana (TSA) para diferentes antibióticos.

b) **Dataset 2:** Presença e ausência de genes com regiões codificantes convertidas em proteínas e genes acessórios ortólogos.

Descrição: Este dataset contém informações sobre a presença e ausência de genes em diferentes isolados. As regiões codificantes foram convertidas em proteínas e genes acessórios ortólogos.


c) **Dataset 3:** Estrutura populacional com clusters baseados na distância dos SNPs (número de locais diferentes) para o genoma central.

Descrição: Este dataset contém dados sobre a estrutura populacional dos isolados, representada por clusters baseados na distância dos SNPs (Single Nucleotide Polymorphisms) para o genoma central.

In [51]:
dataset1 = pd.read_csv("dataset1.csv")
dataset1.head()

,Isolate,Year,CTZ,CTX,AMP,AMX,AMC,TZP,CXM,CET,GEN,TBM,TMP,CIP
0,11657_5#10,2010.0,S,S,S,NaN,S,S,S,S,S,S,S,S
1,11657_5#11,2010.0,S,S,R,NaN,R,S,S,S,S,S,R,R
2,11657_5#12,2010.0,S,S,S,NaN,S,S,S,S,S,S,S,S
3,11657_5#13,2010.0,S,S,R,NaN,R,S,S,S,S,S,S,R
4,11657_5#14,2010.0,S,S,R,NaN,S,S,S,S,S,S,R,S


In [52]:
dataset1.shape

(1936, 14)

- **Número do Isolado (Isolate)**: Número exclusivo para identificar uma cepa específica de bactérias *E. coli*. Portanto, nos referiremos a cada linha do nosso conjunto de dados como um “isolado” de agora em diante.

- **Ano de Isolamento (Year)**: O ano em que uma determinada cepa de bactéria foi separada de seu ambiente natural, daí o nome *isolado*.

- **Antibióticos**: Existem **12 colunas de antibióticos** nomeadas de acordo com sua **abreviatura de 3 letras** adotada pela "Sociedade Britânica de Quimioterapia Antimicrobiana". Página: https://bsac.org.uk/



|Abreviação|Classe: Subclasse|Nome|
|:----------|:--------------|:--------|
|**CTZ**|Beta-lactams: Cephalosporins|Ceftazidime|
|**CTX**|Beta-lactams: Cephalosporins|Cefotaxime |
|**CXM**|Beta-lactams: Cephalosporins|Cefuroxime|
|**CET**|Beta-lactams: Cephalosporins|Cephalothin|
|**AMP**|Beta-lactams: Penicillin|Ampicillin|
|**AMX**|Beta-lactams: Penicillin|Amoxicillin|
|**AMC**|Beta-lactams: Penicillin|Amoxicillin + Clavulanate potassium|
|**TZP**|Beta-lactams: Piperacillin|Tazobactam|
|**GEN**|Aminoglycosides|Gentamicin|
|**TBM**|Aminoglycosides|Tobramycin|
|**TMP**|Antifolate|Trimethoprim|
|**CIP**|Fluoroquinolones|Ciprofloxacin|

**Determinação dos Valores Resistente (R) e Suscetível (S):**

**Teste de suscetibilidade antimicrobiana**: São testes laboratoriais, onde concentrações específicas de um medicamento são testadas para cada isolado de bactéria para determinar se ele é **Resistente (R)**, **Suscetível (S)** ou **Intermediário (I)**.

- **Pontos de interrupção clínicos**: Cada um dos medicamentos tem uma concentração diferente (ponto de interrupção clínico) usada para determinar a resistência ou suscetibilidade de *E. coli* a esse medicamento. Estas concentrações são acordadas pelas agências de saúde pública com base em diferentes fatores e podem variar de ano para ano, dadas as novas descobertas. Eles servem como uma ferramenta de orientação para os médicos prescreverem um antibiótico específico e uma dose apropriada para os pacientes.

- **Os resultados de resistência (R) e suscetível (S)** dos testes laboratoriais foram determinados com base na série de diretrizes do [Comitê Europeu de Testes de Suscetibilidade Antimicrobiana (EUCAST)](https://www.eucast.org/videos_and_online_seminars/english) em 25/01/2017. Para este estudo, os isolados classificados como **Intermediários (I)** foram agrupados com os **Resistentes (R)**.

**Nota**: NaN foi usado para marcar dados faltantes, para aqueles isolados que não foram testados com um medicamento específico.

#### Dataset 2

In [54]:
dataset2 = pd.read_csv("dataset2.csv")
dataset2.head()

,Unnamed: 0,yeiU,yhhS,ybaE,eutR,ibrB,ytfP,aslB,narQ,tolR,...,group_48768,group_48873,group_48916,group_48933,group_48937,group_48958,group_49020,group_49174,group_49253,group_49257
0,11657_5#1,1,1,1,1,1,1,1,1,1,...,0,0,0,0,0,0,0,0,0,0
1,11657_5#10,1,1,1,1,1,1,1,1,1,...,0,0,0,0,0,0,0,0,0,0
2,11657_5#11,1,1,1,1,1,1,1,1,1,...,0,0,0,0,0,0,0,0,0,0
3,11657_5#12,1,1,1,1,1,1,1,1,1,...,0,0,0,0,0,0,0,0,0,0
4,11657_5#13,1,1,1,1,1,1,1,1,1,...,0,0,0,0,0,0,0,0,0,0


In [ ]:
dataset2.shape

(2033, 17199)

In [56]:
# Ajusta o nome da coluna
dataset2 = dataset2.rename(columns = {'Unnamed: 0' : 'Isolate'})

**Genes:** São pedaços de DNA que podem ter tamanhos diferentes. Alguns genes codificam proteínas específicas e sua função é bem compreendida, enquanto a função de outros genes ainda não é bem compreendida. Os genes podem ser geralmente classificados em:

- **Genes Centrais:** Estes são os genes que estão presentes em quase todos os indivíduos para uma espécie bacteriana específica. No nosso caso, são os genes que todos os nossos isolados de *E.coli* têm em comum.

- **Genes Acessórios:** São os genes que podem ser encontrados em um indivíduo, mas não em outro indivíduo da mesma espécie. Esses seriam os genes exclusivos de cada um de nossos isoaldos *E. coli*.

- **Pan-genoma:** Estes são todos os genes possíveis que podem ser encontrados em uma determinada espécie. Ou seja, são todos os genes presentes em nosso isolado *E. coli*.

Consideramos:

- **0** = Ausência do gene
- **1** = Presença do gene.

Como mencionado acima, nem todos os genes têm funções conhecidas e são nomeados, por isso o autor do dataset os separou em 2 conjuntos de genes que foram nomeados de forma diferente:

**1) Regiões Codificantes Conhecidas**: Que foram extraídas das sequências de DNA anotadas e nomeadas de acordo com a proteína que codificam.

No resultado acima os genes codificadores **não contêm a palavra "group_"** no título da coluna.

**2) Regiões Codificantes Desconhecidas**: sequências de DNA que não codificam uma proteína específica, mas que são agrupadas com base em **grupos de genes ortólogos**. Isto significa que estas são sequências onde não sabemos necessariamente a sua função, mas sabemos que elas existem em muitos dos isolados *E. coli*. **Eles recebem um nome que começa com "group_"**.

**Genes Homólogos**: Derivam do mesmo gene ancestral e podem ser:

- **Genes Ortólogos**: São sequências onde não necessariamente sabemos sua função, mas sabemos que elas existem em muitos dos isolados *E. coli*. Para nossos propósitos, estes são todos os isolados *E. coli* que pertencem a um grupo específico.

- **Genes Parálogos**: São sequências geradas através da duplicação de genes no genoma da mesma espécie e são responsáveis por versões alternativas de um determinado gene.

#### Dataset 3

In [69]:
dataset3 = pd.read_csv("dataset3.csv")

In [60]:
dataset3.head()

,Unnamed: 0,cutoff_2,cutoff_3,cutoff_4,cutoff_5,cutoff_6,cutoff_7,cutoff_8,cutoff_9,cutoff_10,...,cutoff_25459,cutoff_25654,cutoff_25772,cutoff_25979,cutoff_26792,cutoff_27119,cutoff_27236,cutoff_27248,cutoff_27690,cutoff_45092
0,11657_5#10,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
1,11657_5#11,1045,1043,1038,1033,1032,1031,1029,1026,1016,...,0,0,0,0,0,0,0,0,0,0
2,11657_5#12,1156,1154,1149,1144,1143,1142,1140,1137,1127,...,0,0,0,0,0,0,0,0,0,0
3,11657_5#13,1267,1265,1260,1255,1254,1253,1251,1248,1238,...,0,0,0,0,0,0,0,0,0,0
4,11657_5#14,1378,1376,1371,1366,1365,1364,1362,1359,1349,...,0,0,0,0,0,0,0,0,0,0


**Estrutura populacional** é definida como a organização da variação genética em uma população ou espécie. É basicamente sobre quão semelhantes (geneticamente) são diferentes isolados. A ideia é que isolados geneticamente semelhantes talvez tenham o mesmo perfil de resistência.

**1) A coluna Isolate** agora é igual à coluna **Isolate** do arquivo csv de metadados.

**2) cutoff_#**: Os nomes deste conjunto de colunas representam os diferentes valores de corte (cutoff) usados para agrupar isolados em diferentes clusters. Por exemplo, o valor de corte de 3 (cutoff_3) significa que agrupamos em um cluster se os isolados tiverem 3 ou menos diferenças (também conhecidos como **SNPs**) entre eles. Ao final do processo cada isolado é classificado em um cluster diferente dependendo do valor de corte. Assim, cada valor de corte deve produzir um conjunto diferente de clusters. Os números abaixo de cada coluna são simplesmente o número de referência para cada cluster.

**NOTA:** **SNP (Polimorfismo de Nucleotídeo Único)** é uma única diferença em um nucleotídeo em um gene específico quando temos um par de isolados sendo comparados. Neste caso, cada isolado no conjunto de dados foi comparado para um **genoma de referência principal EC958**.

Se o valor de corte for menor, podemos esperar que a maioria dos isolados não será agrupada com nenhum outro isolado, formando um número maior de clusters, mas cada cluster seria composto por grupos de tamanhos menores. A soma de todos os membros de cada cluster deve produzir o número total de isolados (ou seja, 1936).

- Por exemplo, para **cutoff_2008** estamos considerando todos os isolados que possuem SNPs de 2008 ou menos. O código abaixo nos mostra que para este valor de corte temos um total de 517 clusters diferentes, cada um deles com um número diferente de isolados. Por exemplo, podemos ver que para o “cluster 222”, temos 221 isolados.

In [70]:
# Alterando a primeira coluna para corresponder aos outros dataframes
dataset3 = dataset3.rename(columns = {'Unnamed: 0' : 'Isolate'})

In [71]:
# Filtrando o cutoff_2008
cutoff_2008 = pd.DataFrame(dataset3['cutoff_2008'].value_counts())

In [72]:
cutoff_2008

,count
cutoff_2008,
222,221
111,208
334,75
134,57
472,45
...,...
452,1
453,1
454,1


Se o valor de corte for alto, podemos esperar que mais isolados sejam agrupados com outros isolados, formando um número menor de clusters, mas a maioria sendo agrupada em um cluster específico. Assim como antes, a soma de todos os membros de cada cluster deve ser igual ao nosso número total de isolados (ou seja, 1936).

- Por exemplo, se aumentarmos o valor de corte para **cutoff_27236**, ​​podemos observar que a maioria dos isolados (1932) cairia no "cluster 0" com apenas alguns isolados dispersos em outros clusters.

In [73]:
# Filtrando o cutoff_27236
cutoff_27236 = pd.DataFrame(dataset3['cutoff_27236'].value_counts())
cutoff_27236

,count
cutoff_27236,
0,1932
1,1
2,1
3,1
4,1


Como não há informações sobre qual valor de corte nos forneceria a melhor maneira de agrupar nossos isolados ou quais membros do grupo estão necessariamente relacionados à resistência (R) ou à suscetibilidade (S), o autor do dataset incluiu 1.072 valores de corte diferentes, cada um com seus próprios grupos.

## Combinação dos Dados



In [81]:
# Os anos em que as cepas foram isoladas
lista_ano = dataset1["Year"].unique()
lista_ano = lista_ano[np.logical_not(np.isnan(lista_ano))]
print(sorted([int(x) for x in lista_ano]))

[1970, 1977, 1994, 1997, 1998, 1999, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2017]


One-hot-encoding na coluna ano:

- **"0"**: O isolamento foi realizado naquele ano.
- **"1"** : O isolamento não foi realizado naquele ano.

In [82]:
# Criando uma variável dummy com one-hot-encoding
dataset1_dummies = pd.get_dummies(dataset1, columns = ["Year"], dummy_na = False)

In [83]:
# Lista de todas as três fontes de dados
df_list = [dataset1_dummies, dataset2, dataset3]

In [84]:
df_combined = reduce(lambda left, right: pd.merge(left, right, on = ['Isolate'], how = 'inner'), df_list)

In [85]:
df_combined.shape

(1936, 18304)

In [86]:
df_combined.sample(5)

,Isolate,CTZ,CTX,AMP,AMX,AMC,TZP,CXM,CET,GEN,...,cutoff_25459,cutoff_25654,cutoff_25772,cutoff_25979,cutoff_26792,cutoff_27119,cutoff_27236,cutoff_27248,cutoff_27690,cutoff_45092
1786,24742_1#227,R,R,R,NaN,NaN,NaN,R,R,S,...,0,0,0,0,0,0,0,0,0,0
234,11657_7#50,S,S,NaN,S,S,S,S,NaN,S,...,0,0,0,0,0,0,0,0,0,0
1486,11791_7#63,S,S,NaN,S,S,S,S,NaN,S,...,0,0,0,0,0,0,0,0,0,0
1796,24742_1#237,R,R,R,NaN,NaN,NaN,R,R,S,...,0,0,0,0,0,0,0,0,0,0
1865,24742_1#329,R,S,R,NaN,NaN,NaN,R,R,S,...,0,0,0,0,0,0,0,0,0,0


- Observe que o número de linhas corresponde corretamente ao número de isolados, resultando em um total de 1.936 linhas como nos metadados.

- E temos várias colunas que incluem:

     - **1 coluna de número de isolado** com as tags exclusivas para cada um de nossos isolados.
     - **12 rótulos**, um para cada medicamento para o qual tentaremos fazer previsões.
     - **18291 recursos** que usaremos para fazer previsões para os rótulos.
 
Total de colunas: 18304

In [87]:
# Exporta para CSV
df_combined.to_csv("dataset_final.csv", index = False)